File: src/medium/MinSubArrayLen.go
package main

import "math"

/*给定一个含有 n 个正整数的数组和一个正整数 target 。
找出该数组中满足其和 ≥ target 的长度最小的 连续子数组 [numsl, numsl+1, ..., numsr-1, numsr]
，并返回其长度。如果不存在符合条件的子数组，返回 0 。
示例 1：
输入：target = 7, nums = [2,3,1,2,4,3]
输出：2
解释：子数组 [4,3] 是该条件下的长度最小的子数组。*/

func main() {

}
func minSubArrayLen(s int, nums []int) int {
	//暴力破解法
	n := len(nums)
	if n == 0 {
		return 0
	}
	ans := math.MaxInt32
	for i := 0; i < n; i++ {
		sum := 0
		for j := i; j < n; j++ {
			sum += nums[j]
			if sum >= s {
				ans = min(ans, j-i+1)
				break
			}
		}
	}
	if ans == math.MaxInt32 {
		return 0
	}
	return ans
}

func min(x, y int) int {
	if x < y {
		return x
	}
	return y
}

func minSubArrayLen2(s int, nums []int) int {
	//滑动数组法
	n := len(nums)
	if n == 0 {
		return 0
	}
	ans := math.MaxInt32
	start, end := 0, 0
	sum := 0
	for end < n {
		sum += nums[end]
		for sum >= s {
			ans = min(ans, end-start+1)
			sum -= nums[start]
			start++
		}
		end++
	}
	if ans == math.MaxInt32 {
		return 0
	}
	return ans
}

File: src/medium/MinSubArrayLen.java
package medium;

/**
 * @author siianchan
 * @since 2024/5/23 14:24
 */
public class MinSubArrayLen {
    /**
     * 给定一个含有 n 个正整数的数组和一个正整数 target 。
     * 找出该数组中满足其总和大于等于 target 的长度最小的 连续
     * 子数组
     * [numsl, numsl+1, ..., numsr-1, numsr] ，并返回其长度。如果不存在符合条件的子数组，返回 0 。
     */
    public static void main(String[] args) {
        byte a=(byte) Integer.MAX_VALUE;
        System.out.println(a);
    }

    public int minSubArrayLen(int target, int[] nums) {
        int ret = Integer.MAX_VALUE;
        int left = 0;
        int now = 0;
        for (int right = 0; right < nums.length; right++) {
            now += nums[right];
            //滑动数组，当满足条件时，滑动左边下标，看是否满足
            while (now - nums[left] >= target) {
                now -= nums[left];
                left++;
            }
            if (now >= target) {
                ret = Math.min(ret, right - left + 1);
            }
        }
        return ret <= nums.length ? ret : 0;
    }
}
